File: BackUpForABook/UIControls/Common/DSSlider.m
//
//  DSSlider.m
//  ATreeRingMap
//
//  Created by Hui Wang on 5/29/14.
//  Copyright (c) 2014 AnnieLyticx. All rights reserved.
//

#import "DSSlider.h"

@interface DSSlider (PrivateMethods)
// paint slider bar
-(void)paintBar:(CGContextRef)ctx;
// paint bullet arrow
-(void)paintArrows:(CGContextRef)ctx;
// test functions
-(void)paintBarTest:(CGContextRef)ctx;
@end

@implementation DSSlider

-(id)init
{
    self = [super init];
    
    _length = 30;
    _height = 8;
    _space = 8;
    
    _faceColor = [UIColor colorWithRed:0.0 green:.45 blue:.94 alpha:.8];
    
    return self;
}

// paint the slider
-(void)paint:(CGContextRef)ctx
{
    [self paintBar:ctx];
    [self paintArrows:ctx];
}

// paint slider bar
-(void)paintBar:(CGContextRef)ctx
{
    CGContextSaveGState(ctx);
    
    CGRect slider = CGRectMake(-_length * .5, -_height * .5, _length, _height);
    
    UIBezierPath *aPath = [UIBezierPath bezierPathWithRoundedRect:slider cornerRadius:_height / 2.0];
    CGContextAddPath(ctx, aPath.CGPath);
    
    // (1) fill the slider area
    CGContextSetFillColorWithColor(ctx, _faceColor.CGColor);
    CGContextAddPath(ctx, aPath.CGPath);
    CGContextFillPath(ctx);
    
    // (2) add a highlight over the bar
    CGRect highlight = CGRectMake(-_length * .5 + _height * .5, 0.0, _length - _height, _height * .5);
    UIBezierPath *aHighlightPath = [UIBezierPath bezierPathWithRoundedRect:highlight cornerRadius:highlight.size.height * 1.0 / 2.0];
    CGContextAddPath(ctx, aHighlightPath.CGPath);
    CGContextSetFillColorWithColor(ctx, [UIColor colorWithWhite:1.0 alpha:.4].CGColor);
    CGContextFillPath(ctx);
    
    // (3) inner shadow
    CGContextSetShadowWithColor(ctx, CGSizeMake(0, 2.0), 3.0, [UIColor grayColor].CGColor);
    CGContextAddPath(ctx, aPath.CGPath);
    CGContextSetStrokeColorWithColor(ctx, [UIColor grayColor].CGColor);
    CGContextStrokePath(ctx);
    
    // (4) outline the slider
    CGContextAddPath(ctx, aPath.CGPath);
    CGContextSetStrokeColorWithColor(ctx, [UIColor darkGrayColor].CGColor);
    CGContextSetLineWidth(ctx, .5);
    CGContextStrokePath(ctx);
    
    CGContextRestoreGState(ctx);
}

// paint bullet arrow
-(void)paintArrows:(CGContextRef)ctx
{
    // paint the left and right bullet arrow
    int arrowSize = _height * .5;
    
    CGPoint leftArrowPosition = CGPointMake(0 - _length * .5 - _space, .0);
    CGPoint rightArrowPosition = CGPointMake(0 + _length * .5 + _space, .0);
    
    // left arrow
    UIBezierPath *arrowPath = [UIBezierPath bezierPath];
    [arrowPath moveToPoint:CGPointMake(leftArrowPosition.x + arrowSize, leftArrowPosition.y - arrowSize)];
    [arrowPath addLineToPoint:leftArrowPosition];
    [arrowPath addLineToPoint:CGPointMake(leftArrowPosition.x + arrowSize, leftArrowPosition.y + arrowSize)];
    CGContextAddPath(ctx, arrowPath.CGPath);
    
    CGContextSetShadowWithColor(ctx, CGSizeMake(0, 2.0), 3.0, [UIColor grayColor].CGColor);
    CGContextSetStrokeColorWithColor(ctx, _faceColor.CGColor);
    CGContextSetLineWidth(ctx, 2.0);
    CGContextStrokePath(ctx);
    
    // right arrow
    [arrowPath removeAllPoints];
    
    [arrowPath moveToPoint:CGPointMake(rightArrowPosition.x - arrowSize, rightArrowPosition.y - arrowSize)];
    [arrowPath addLineToPoint:rightArrowPosition];
    [arrowPath addLineToPoint:CGPointMake(rightArrowPosition.x - arrowSize, rightArrowPosition.y + arrowSize)];
    CGContextAddPath(ctx, arrowPath.CGPath);
    
    CGContextStrokePath(ctx);
}

// test functions
-(void)paintBarTest:(CGContextRef)ctx
{
    CGContextSaveGState(ctx);
    
    CGRect slider = CGRectMake(-_length * .5, -_height * .5, _length, _height);
    
    UIBezierPath *aPath = [UIBezierPath bezierPathWithRoundedRect:slider cornerRadius:_height / 2.0];
    CGContextAddPath(ctx, aPath.CGPath);
    
    // (1) fill the slider area
    CGContextSetFillColorWithColor(ctx, _faceColor.CGColor);
    CGContextAddPath(ctx, aPath.CGPath);
    CGContextFillPath(ctx);
    
    // (2) add a highlight over the bar
    CGRect highlight = CGRectMake(-_length * .5 + _height * .5, 0.0, _length - _height, _height * .5);
    UIBezierPath *aHighlightPath = [UIBezierPath bezierPathWithRoundedRect:highlight cornerRadius:highlight.size.height * 1.0 / 2.0];
    CGContextAddPath(ctx, aHighlightPath.CGPath);
    CGContextSetFillColorWithColor(ctx, [UIColor colorWithWhite:1.0 alpha:.4].CGColor);
    CGContextFillPath(ctx);
    
    // (3) inner shadow
    CGContextSetShadowWithColor(ctx, CGSizeMake(0, 2.0), 3.0, [UIColor grayColor].CGColor);
    CGContextAddPath(ctx, aPath.CGPath);
    CGContextSetStrokeColorWithColor(ctx, [UIColor grayColor].CGColor);
    CGContextStrokePath(ctx);
    
    // (4) outline the slider
    CGContextAddPath(ctx, aPath.CGPath);
    CGContextSetStrokeColorWithColor(ctx, [UIColor darkGrayColor].CGColor);
    CGContextSetLineWidth(ctx, .5);
    CGContextStrokePath(ctx);
    
    CGContextRestoreGState(ctx);
}

@end

File: BackUpForABook/UIControls/TRClasses_swift/DSSlider.swift
//
//  DSSlider.swift
//  TreeRingMap_Swift
//
//  Created by dingf on 17/3/16.
//  Copyright © 2017年 MH.dingf. All rights reserved.
//

import UIKit

class DSSlider: NSObject {
    var length: Float?
    var height: Float?
    var space: Int?
    var faceColor: UIColor?
    
    override init() {
        super.init()
        length = 30
        height = 8
        space = 8
        faceColor = UIColor(red: 0, green: 0.45, blue: 0.94, alpha: 0.8)
    }
    
    func paint(_ ctx: CGContext){
        paintBar(ctx)
        paintArrows(ctx)
    }
    
    func paintBar(_ ctx: CGContext){
        ctx.saveGState()
        let slider = CGRect(x: CGFloat(-length! * 0.5), y: CGFloat(-height! * 0.5), width: CGFloat(length!), height: CGFloat(height!))
        let aPath = UIBezierPath(roundedRect: slider, cornerRadius: CGFloat(height! / 2.0))
        ctx.addPath(aPath.cgPath)
        // fill the slider area
        ctx.setFillColor(faceColor!.cgColor)
        ctx.addPath(aPath.cgPath)
        ctx.fillPath()
        // add a highlight over the bar
        let highlight = CGRect(x: CGFloat(-length! * 0.5 + height! * 0.5), y: 0.0, width: CGFloat(length! - height!), height: CGFloat(height! * 0.5))
        let aHighlightPath = UIBezierPath(roundedRect: highlight, cornerRadius: highlight.size.height * 1.0 / 2.0)
        ctx.addPath(aHighlightPath.cgPath)
        ctx.setFillColor(UIColor(white: 1.0, alpha: 0.4).cgColor)
        ctx.fillPath()
        // inner shadow
        ctx.setShadow(offset: CGSize.init(width: 0, height: 2.0), blur: 3.0, color: UIColor.gray.cgColor)
        ctx.addPath(aPath.cgPath)
        ctx.setStrokeColor(UIColor.gray.cgColor)
        ctx.strokePath()
        // outline the slider
        ctx.addPath(aPath.cgPath)
        ctx.setStrokeColor(UIColor.darkGray.cgColor)
        ctx.setLineWidth(0.5)
        ctx.strokePath()
        ctx.restoreGState()
    }
    
    func paintArrows(_ ctx: CGContext){
        let arrowSize = height! * 0.5
        let leftArrowPositon = CGPoint(x: CGFloat(0 - length! * 0.5 - Float(space!)), y: 0.0)
        let rightArrowPosition = CGPoint(x: CGFloat(0 + length! * 0.5 + Float(space!)), y: 0.0)
        // left arrow
        let arrowPath = UIBezierPath()
        arrowPath.move(to: CGPoint(x: leftArrowPositon.x + CGFloat(arrowSize), y: leftArrowPositon.y - CGFloat(arrowSize)))
        arrowPath.addLine(to: leftArrowPositon)
        arrowPath.addLine(to: CGPoint(x: leftArrowPositon.x + CGFloat(arrowSize), y: leftArrowPositon.y + CGFloat(arrowSize)))
        ctx.addPath(arrowPath.cgPath)
        ctx.setShadow(offset: CGSize.init(width: 0, height: 2.0), blur: 3.0, color: UIColor.gray.cgColor)
        ctx.setStrokeColor(faceColor!.cgColor)
        ctx.setLineWidth(2.0)
        ctx.strokePath()
        
        // right arrow
        arrowPath.removeAllPoints()
        arrowPath.move(to: CGPoint(x: rightArrowPosition.x - CGFloat(arrowSize), y: rightArrowPosition.y - CGFloat(arrowSize)))
        arrowPath.addLine(to: rightArrowPosition)
        arrowPath.addLine(to: CGPoint(x: rightArrowPosition.x - CGFloat(arrowSize), y: rightArrowPosition.y + CGFloat(arrowSize)))
        ctx.addPath(arrowPath.cgPath)
        ctx.strokePath()
    }

}
